firebase-vscode/webviews/globals/extension-state.tsx

```diff
@@ -0,0 +1,64 @@
+import React, { createContext, ReactNode, useContext, useEffect } from "react";
+import { broker } from "./html-broker";
+import { signal, computed } from "@preact/signals-react";
+import { User } from "../types/auth";
+
+export enum Environment {
+  UNESPECIFIED,
+  VSC,
+  IDX,
+}
+
+function createExtensionState() {
+  const environment = signal(Environment.UNESPECIFIED);
+  const users = signal<User[]>([]);
+  const selectedUserEmail = signal("");
+  const projectId = signal("");
+
+  const selectedUser = computed(() =>
+    users.value.find((user) => user.email === selectedUserEmail.value)
+  );
+
+  return { environment, users, projectId, selectedUserEmail, selectedUser };
+}
+
+const ExtensionState =
+  createContext<ReturnType<typeof createExtensionState>>(null);
+
+export function ExtensionStateProvider({
+  children,
+}: {
+  children: ReactNode;
+}): JSX.Element {
+  const state = createExtensionState();
```

Comment:
Will this run every time the component re-renders? Would it create a duplicate ExtensionState?

Reply:
Technically very much yes, but that's why it's all the way up at the top of the React tree. I'll add some comments.